code/go/0chain.net/core/common/lock.go

```diff
@@ -0,0 +1,78 @@
+package common
+
+import (
+	"sync"
+)
+
+type Lock struct {
+	key string
+	// stale signals if lock is deleted from map. if true, discard use of this lock and get fresh lock
+	stale      bool
+	actualLock *sync.Mutex
+	// count how many process requires this lock
+	count   int32
+	countMu *sync.Mutex
+
+	// Parent Map
+	pMap *MapLocker
+}
+
+// Lock Acquire lock
+func (l *Lock) Lock() {
+	for {
+		l.actualLock.Lock()
+		if l.stale {
+			l = l.pMap.GetLock(l.key)
+			continue
+		}
+		break
+	}
+}
+
+func (l *Lock) Unlock() {
+	l.actualLock.Unlock()
+	l.countMu.Lock()
+	l.count--
+	if l.count == 0 {
+		l.stale = true
+		l.pMap.m.Delete(l.key)
+	}
+	l.countMu.Unlock()
+}
+
+type MapLocker struct {
+	m *sync.Map
+	// Reduce updating lock with GetLock call with same key
+	mu *sync.Mutex
+}
+
+func (m *MapLocker) GetLock(key string) (l *Lock) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	valueI, ok := m.m.Load(key)
```

Comment:
`lock.Unlock()` deletes the key from the `MapLocker` without the protect of the `m.mu` lock, imagine the situation that we acquired the `m.mu.Lock()` and load the lock from the map here, then in another goroutine, this lock was unlocked, and it's count reduce to 0, then it would be marked as `stale` and removed from the `MapLocker`. The goroutine comes back here again and continue, the lock's `count` updated, and returned to user. However, it was not in the MapLocker anymore, we could have thread safe issue if user try to get the lock of the same key again in a new goroutine, as the MapLocker will give a new one, thus we will try to lock the same resources with different locks, which obviously, would not work as expected.

Reply:
>The goroutine comes back here again and continue, the lock's `count` updated, and returned to user. However, it was not in the MapLocker anymore,

@peterlimg I don't see how lock's `count` will get updated if it does not exist in the `mapLocker`. Only one goroutine will have stale lock but it will check against staleness and calls `GetLock` again. It will get new lock with `count=1`. `GetLock` is protected by `m.mu.Lock` and further `l.countMu.Lock` will protect count update.

Reply:
The lock was acquired by another thread. Let me explain in details and see the case below:

1. A gorountine called the `GetLock()` and get a lock `A`, it will use `A.Lock()` then `A.Unlock()` to release the lock. So suppose we called `A.Lock()` but hadn't called the `A.Unlock()`
2. Another gorountine come to the `mapLocker.GetLock()` function as show above, this goroutine will get the lock `A`  as well if the same key was used. This goroutine will get the lock `A` and suppose it come  to the code in line `52` where it just loaded the `valueI` from the map, then it go back to the previous goroutine.
3. The previosus goroutine called the `A.Unlock()` as show below, apparently, the `count` would be reduced to `0` and  then it will delete it self from the map locker:
   ```go
    func (l *Lock) Unlock() {
	    l.actualLock.Unlock()
	    l.countMu.Lock()
	    l.count--
	    if l.count == 0 {
		    l.stale = true
		    l.pMap.m.Delete(l.key)
	    }
	    l.countMu.Unlock()
}
   ```
4. Come back to the newer goroutine, it continue as the code show above, this lock `A` will be returned to the user, but it does not exist in the map anymore.

Reply:
@peterlimg While unlocking, a goroutine also achieves `l.countMu.Lock()` and reduce count by 1. In `GetLock` function it should also get `l.countMu.Lock` to update count.

Suppose `A.Unlock()` reduce count to 0, made itself `stale` and removed it from `mapLocker`. When another goroutine which had got same lock when tries to achieve `A.Lock` it also checks if it is `stale` in line 24. This `staleness` should ensure that there are no two locks for same key

Reply:
@lpoli yes, you're right, the stale checking should be able to ensure no two locks for the same key. So this looks good to me as long as the previous `locker pointer` assignment issue is fixed.